core/lib/Cardano/Address/Script.hs

```diff
@@ -266,35 +301,75 @@ foldScript fn zero = \case
 -- | Validate a 'Script', semantically
 --
 -- @since 3.0.0
-validateScript :: Script -> Either ErrValidateScript ()
-validateScript = \case
-    RequireSignatureOf (KeyHash bytes) -> do
-        when (BS.length bytes /= credentialHashSize) $ Left WrongKeyHash
+validateScript :: Script KeyHash -> Either ErrValidateScript ()
+validateScript script = do
+    let validateKeyHash (KeyHash bytes) =
+            when (BS.length bytes /= credentialHashSize) $ Left WrongKeyHash
+    validateScript' validateKeyHash script
+
+validateScript'
+    :: Eq elem
+    => (elem -> Either ErrValidateScript ())
+    -> Script elem
+    -> Either ErrValidateScript ()
+validateScript' validateRequireSignatureOf = \case
+    RequireSignatureOf element ->
+        validateRequireSignatureOf element
 
     RequireAllOf script -> do
         when (L.null script) $ Left EmptyList
         when (hasDuplicate script) $ Left DuplicateSignatures
-        traverse_ validateScript script
+        when (invalidTimelocks script) $ Left InvalidTimelocks
+        traverse_ (validateScript' validateRequireSignatureOf) script
 
     RequireAnyOf script -> do
         when (L.null script) $ Left EmptyList
         when (hasDuplicate script) $ Left DuplicateSignatures
-        traverse_ validateScript script
+        when (invalidTimelocks script) $ Left InvalidTimelocks
+        traverse_ (validateScript' validateRequireSignatureOf) script
 
     RequireSomeOf m script -> do
         when (m == 0) $ Left MZero
         when (length script < fromIntegral m) $ Left ListTooSmall
         when (hasDuplicate script) $ Left DuplicateSignatures
-        traverse_ validateScript script
+        when (invalidTimelocks script) $ Left InvalidTimelocks
+        traverse_ (validateScript' validateRequireSignatureOf) script
 
     ActiveFromSlot _ -> pure ()
 
     ActiveUntilSlot _ -> pure ()
   where
-    hasDuplicate xs = do
+    hasDuplicate xs =
         length sigs /= length (L.nub sigs)
       where
         sigs = [ sig | RequireSignatureOf sig <- xs ]
+    invalidTimelocks xs =
+        let hasTimelocks = \case
+                ActiveFromSlot _ -> True
+                ActiveUntilSlot _ -> True
+                _ -> False
+        in case filter hasTimelocks xs of
+            [] -> False
+            [ActiveFromSlot s1, ActiveUntilSlot s2] -> s2 <= s1
+            [ActiveUntilSlot s2, ActiveFromSlot s1] -> s2 <= s1
+            [ActiveFromSlot _] -> False
+            [ActiveUntilSlot _] -> False
+            _ -> True
```

Comment:
Arguably, `[ActiveFromSlot t1, ActiveFromSlot t2]` is va valid script (although one of the two clauses is redundant). I don't think we should do any particular validation with regards to scripts yet. It gives a false sense of safeness whereas there are still plenty of ways to shoot oneself in the foot.

Reply:
I agree, one can construct any ScriptTemplate value. but if one wants to validate it and somewhat get feedback on what are reasonable things to use then can refer to `validateScriptTemplate`.  I do think using several ActiveFromSlot in a given nesting should be discouraged as it can give rise to problems one day...